gusb/gusb-umockdev-test.c

```diff
@@ -0,0 +1,211 @@
+		"A: speed=480\\n\n"
+		/* descriptor from a Canon PowerShot SX200; VID 04a9 PID 31c0 */
+		"H: descriptors="
+		  "1201000200000040a904c03102000102"
+		  "030109022700010100c0010904000003"
+		  "06010100070581020002000705020200"
+		  "020007058303080009\n",
+		NULL);
+}
+
+static void
+test_ctx_enumerate (UMockdevTestbedFixture *fixture, UNUSED_DATA)
+{
+	g_autoptr(GPtrArray) devices = NULL;
+
+	test_fixture_add_canon (fixture);
+
+	g_usb_context_enumerate (fixture->ctx);
+
+	devices = g_usb_context_get_devices (fixture->ctx);
+	g_assert_cmpint (devices->len, ==, 1);
+}
+
+static void
+count_hotplug_event_cb (GUsbContext *context, GUsbDevice *device, gpointer user_data)
+{
+	int *counter = user_data;
+
+	*counter += 1;
+}
+
+static void
+test_ctx_hotplug (UMockdevTestbedFixture *fixture, UNUSED_DATA)
+{
+	g_autoptr(GPtrArray) devices = NULL;
+	gint events = 0;
+
+	g_signal_connect (fixture->ctx, "device-added", G_CALLBACK (count_hotplug_event_cb), &events);
+
+	g_usb_context_enumerate (fixture->ctx);
+
+	devices = g_usb_context_get_devices (fixture->ctx);
+	g_assert_cmpint (devices->len, ==, 0);
+	g_assert_cmpint (events, ==, 0);
+	g_clear_pointer (&devices, g_ptr_array_unref);
+
+	test_fixture_add_canon (fixture);
+	/* Ensure the event was processed by helper thread. */
+	g_usleep (G_USEC_PER_SEC / 2);
+
+	/* Still not returned (and no event fired). */
+	devices = g_usb_context_get_devices (fixture->ctx);
+	g_assert_cmpint (devices->len, ==, 0);
+	g_assert_cmpint (events, ==, 0);
+	g_clear_pointer (&devices, g_ptr_array_unref);
+
+	/* Run mainloop, which causes the event to be processed. */
+	while (g_main_context_iteration (NULL, FALSE)) { }
+
+	devices = g_usb_context_get_devices (fixture->ctx);
+	g_assert_cmpint (events, ==, 1);
+	g_assert_cmpint (devices->len, ==, 1);
+	g_clear_pointer (&devices, g_ptr_array_unref);
+}
+
+static void
+test_ctx_hotplug_dispose (UMockdevTestbedFixture *fixture, UNUSED_DATA)
+{
+	g_autoptr(GPtrArray) devices = NULL;
+	gint events = 0;
+
+	g_signal_connect (fixture->ctx, "device-added", G_CALLBACK (count_hotplug_event_cb), &events);
+
+	g_usb_context_enumerate (fixture->ctx);
+	devices = g_usb_context_get_devices (fixture->ctx);
+	g_assert_cmpint (devices->len, ==, 0);
+	g_assert_cmpint (events, ==, 0);
+	g_clear_pointer (&devices, g_ptr_array_unref);
+
+	test_fixture_add_canon (fixture);
+	/* Ensure the event was processed by helper thread. */
+	g_usleep (G_USEC_PER_SEC / 2);
+
+	/* Still not returned (and no event fired). */
+	g_usb_context_enumerate (fixture->ctx);
+	devices = g_usb_context_get_devices (fixture->ctx);
+	g_assert_cmpint (devices->len, ==, 0);
+	g_assert_cmpint (events, ==, 0);
+	g_clear_pointer (&devices, g_ptr_array_unref);
+
+	/* The idle handler is pending, we dispose our context reference */
+	g_object_run_dispose (G_OBJECT (fixture->ctx));
+
+	/* Run mainloop, which causes the event to be processed. */
+	while (g_main_context_iteration (NULL, FALSE)) { }
```

Comment:
This is where the crash happens without the fix.

 * `test_fixture_add_canon` generates the hotplug event
 * `g_usb_context_hotplug_cb` is called in libusb event thread (the sleep ensures this happens)
 * We dispose the GUsbContext
 * mainloop runs: `g_usb_context_idle_hotplug_cb` runs and calls `g_usb_context_add_device`

Reply:
Yeah, it doesn't prevent dispatching, it should prevent issues.

Your solution looks better to me, just one typo. I'll file a new MR for the tests (or you can just cherry-pick them).